### PR TITLE
Add SPM support with XCF

### PR DESCRIPTION
**Pull request by julepka**
Introducing Swift Package Manager support with xcframework.

Details:
- Created a script `create_xcframewok.sh` to build an xcframwork for Swift for iOS, iOSSimulator, MacOS. Basically, took it from AppSpector tutorial. The xcframework is created in the `build` folder. When releasing a new version, we'll need to add the generated xcframework to the release assets to have a direct link to it. Related documentation will be updated.
- Added `Package.swift` file to support SPM. It should be placed in the root of the repo. It contains a link to the xcframework. When releasing a new version, the link should be updated with a proper version (future version you are going to release, pretty similar to carthage). So, when you create a release/tag on a GitHub you can reference a corresponding commit. 
- Example projects will be added in the upcoming version together with updated OpenSSL version.
- I had to remove OpenSSL from SPM inside Themis.xcodeproj because of the signature issues. It failed our Carthage tests. It seems to be resolved with Xcode 12.5 that is not available for Github actions yet.

## Checklist

- [x] Change is covered by automated tests – will be covered separately, added to our TODO list
- [x] The [coding guidelines] are followed
- [x] Example projects and code samples are up-to-date (in case of API changes) – will be covered separatly
- [x] Changelog is updated (in case of notable or breaking changes) – TODO when links are updated

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: Themis.xcodeproj/project.pbxproj
Comment (vixentael) [inline Themis.xcodeproj/project.pbxproj:1897]:
```suggestion
				minimumVersion = 1.1.108203;
```

TODO: re-define version after merging https://github.com/cossacklabs/openssl-apple/pull/22

Comment (julepka) [reply Themis.xcodeproj/project.pbxproj:1897]: yeah, and the link will point to the cosacklabs repo.

**Comment by ilammy** [comment]
> Also, I should probably move the create_xcframeworks.sh script to some folder. I believe the scripts one should be good. Correct me if I'm wrong.

I think that `scripts/create_xcframeworks.sh` would be the right location, yes.

File: Package.swift
Comment (ilammy) [inline Package.swift:13]: At first I wanted to ask, “Where is `cl-openssl`?“ But then I realized that it's not here because themis.xcframework will be statically linked to openssl.xcframework, hence users of themis.xcframework don't need to install shared OpenSSL.

File: Package.swift
Comment (ilammy) [inline Package.swift:16]: I guess you could replace this with more informative instructions for maintainers, like in CLOpenSSL repo.

File: Themis.xcodeproj/project.pbxproj
Comment (ilammy) [inline Themis.xcodeproj/project.pbxproj:1896]: Given that versioning here is mostly nonsensical, I believe it would be more reasonable to pin the “exact” version. OpenSSL *does not* use semver, the second digit is their major version and SPM will think it's the first. For example, 1.0.2 => 1.1 has introduced a lot of breakage.

It's not much of a problem given that a) users will not be affected, only maintainers, and b) we maintain and control CLOpenSSL releases as well. But if there is a way to prevent nasty surprises, I'd take it.

You can't shoehorn OpenSSL versioning scheme into semver. (Well, actually *you can*, but using 101.1.10802 for 1.1.1h-2 is hardly elegant.)

File: create_xcframeworks.sh
Comment (ilammy) [inline create_xcframeworks.sh:4]: It would be nice to add a common header to the script:

```bash
#!/usr/bin/env bash
#
# Shebang to explicitly use bash and not break on macOS boxes with zsh as default shell.
#
# After that you usually describe what this script is for and how to use it. In particular,
# what it expects the current directory to be. (That's more important if it's it "scripts".)

set -eu # common flags to ensure that the shell does not ignore failures
```

I'd also put all temporary things into the `build` directory (overridable by `BUILD_PATH` environment variable). Native build system places its stuff there, that directory is already in gitignore, etc.

File: create_xcframeworks.sh
Comment (ilammy) [inline create_xcframeworks.sh:12]: It's a bit confusing to set both variables on the same line.

File: create_xcframeworks.sh
Comment (ilammy) [inline create_xcframeworks.sh:46]: Czechsum 😂

**Review by ilammy — COMMENTED** [review]
Overall, looks good! Thank you, @julepka, for your brain tissue donation to the Apple Gods. I'm sorry you had to do this.

My questions from https://github.com/cossacklabs/openssl-apple/pull/22 applies here as well: regarding the static/dynamic divide. Just in case Themis would need to become a static framework, or switch to shared dynamic OpenSSL, it's interesting to know how painful that would be.

Also, there are some new for Themis future specifically:

- Since there is no precedent of attaching binaries to Themis releases, how that would play out in the general release flow?
- Given that CLOpenSSL serves as a model example of making binary-only releases, what would become to CocoaPods and Carthage publications of Themis? Will they eventually migrate to using the same binaries as SPM, or continue building Themis from source?

**Comment by ilammy** [comment]
Oh, and one more thing: it would be cool to exercise the script on CI, just to be sure it will work in the time of need.

**Comment by julepka** [comment]
@ilammy 
> My questions from [cossacklabs/openssl-apple#22](https://github.com/cossacklabs/openssl-apple/pull/22) applies here as well: regarding the static/dynamic divide. Just in case Themis would need to become a static framework, or switch to shared dynamic OpenSSL, it's interesting to know how painful that would be.

The only thing I can say, it is very unpredictable. I've tried the most straightforward and recommended solution. I hope to make it work as an MVP for out XCF and SPM support. Then. we'll see what we can do with the linking type.

> Since there is no precedent of attaching binaries to Themis releases, how that would play out in the general release flow?

Yeah, looks like we'll need to describe XCF flow in our releasing checklist. In a perfect world, CI will do everything for us, but for now, we will build and attach XCF manually. We will make XCF work and then improve whats missing. 

> Given that CLOpenSSL serves as a model example of making binary-only releases, what would become to CocoaPods and Carthage publications of Themis? Will they eventually migrate to using the same binaries as SPM, or continue building Themis from source?

It may be possible, but I can't say for sure right now.

File: Package.swift
Comment (ilammy) [inline Package.swift]: Thanks for maintenance instructions!

I'd also drop the `$` from `$swift package...` to make this line a bit easier to copy-paste.

File: scripts/create_xcframeworks.sh
Comment (ilammy) [inline scripts/create_xcframeworks.sh:6]: This was more of a placeholder text to be replaced rather than a literal suggestion 😄

File: scripts/create_xcframeworks.sh
Comment (ilammy) [inline scripts/create_xcframeworks.sh:11]: It would be nice to use
```bash
BUILD_PATH=${BUILD_PATH:-build}
```
so that the *default* value for this variable is `build` but it's possible to override it via environment, just like the Makefile allows it.

File: scripts/create_xcframeworks.sh
Comment (ilammy) [inline scripts/create_xcframeworks.sh:14]: Hm... I'd argue it would be easier to just expect the script to be invoked as `scripts/create_xcframeworks.sh` from the project root. That's more natural for CI pipelines, etc. That's what relative path in `BUILD_PATH` expects too.

File: scripts/create_xcframeworks.sh
Comment (ilammy) [inline scripts/create_xcframeworks.sh:20]: If you use `-p` flag, it wouldn't be necessary to create intermediate directories:
```bash
mkdir -p $output_dir/archives
mkdir -p $output_dir/iphoneos
mkdir -p $output_dir/macosx
```

File: scripts/create_xcframeworks.sh
Comment (julepka) [reply scripts/create_xcframeworks.sh:6]: Oh 😄

File: scripts/create_xcframeworks.sh
Comment (julepka) [reply scripts/create_xcframeworks.sh:11]: got it, thanks

**Comment by julepka** [comment]
I had to remove cl-openssl SPM packege from the Themis.xcodeproj because Xcode 12 has signing issues if you have a complex structure if binary frameworks dependencies. https://developer.apple.com/forums/thread/656367 

It looks like the issue is fixed in Xcode 12.5 but github actions do not support it yet.

**Comment by ilammy** [comment]
> I had to remove cl-openssl SPM packege from the Themis.xcodeproj

Hm... So it still depends on that package, but hides it so that Xcode stops eating glue? 🤔

> It looks like the issue is fixed in Xcode 12.5 but github actions do not support it yet.

Well, they seem to support it in their macOS 11 image, but that one has been in “private preview mode” for quite a while.

Now that you've mentioned it, I'm kinda sad because Xcode 12.5 *requires* macOS 11. I'm in for a painful upgrade.

**Comment by vixentael** [comment]
> Now that you've mentioned it, I'm kinda sad because Xcode 12.5 requires macOS 11. I'm in for a painful upgrade.

all of us... all of us 💀

File: CHANGELOG.md
Comment (vixentael) [inline CHANGELOG.md:166]:
```suggestion
- `themis` now supports SPM, its installation and usage are very straightforward, just add `themis` as SPM dependency.
```

**Review by vixentael — APPROVED** [review]

**Pull request merged**
